File: Security/caesar_cipher/caesar_cipher.kt
//Uppercase A starts at 65 | Ends at 90
val upper = 65..90
//Lowercase a starts at 97 | Ends at 122
val lower = 97..122

//Encrypt/Decrypt are just shifting by opposite directions (It works with negatives)
fun main(){
    val original = "Hello World"
    val encrypted = original.shiftRight(2)
    val decrypted = encrypted.shiftLeft(2)
    println(encrypted) //Outputs Jgnnq Yqtnf
    println(decrypted) //Outputs Hello World
}

//It shifts the Ascii value, and uses modulus to keep it within bounds
fun String.shiftRight(amount:Int) = map { c ->
    val d = c.code

    when {
        upper.contains(d) -> {
            val shift = (d + amount) % upper.last
            when {
                shift < upper.first -> (shift + upper.first).toChar()
                else -> shift.toChar()
            }
        }
        lower.contains(d) -> {
            val shift = (d + amount) % lower.last
            when {
                shift < lower.first -> (shift + lower.first).toChar()
                else -> shift.toChar()
            }
        }
        else -> c
    }
}.joinToString("")

fun String.shiftLeft(amount:Int) = map { c ->
    val d = c.code

    when {
        upper.contains(d) -> if (d - amount < upper.first) (upper.last + upper.first - amount - d).toChar() else (d - amount).toChar()
        lower.contains(d) -> if (d - amount < lower.first) (lower.last + lower.first - amount - d).toChar() else (d - amount).toChar()
        else -> c
    }
}.joinToString("")

File: Security/caesar_cipher/caesar_cipher.rs
fn main() {
    let original = "Hello World";
    let encrypted = original.shift_right(2);
    let decrypted = &encrypted.shift_left(2);
    println!("{}", encrypted); // Outputs Jgnnq Yqtnf
    println!("{}", decrypted); // Outputs Hello World
}

trait StrExt {
    fn shift_right(&self, amount: usize) -> String;
    fn shift_left(&self, amount: usize) -> String;
}

#[rustfmt::skip]
impl StrExt for str {
    fn shift_right(&self, amount: usize) -> String {
        self.chars()
            .map(|alph| {
                std::char::from_u32(match alph {
                    _ if alph.is_ascii_uppercase() => (65..= 90).cycle().nth(alph as usize - 65 + amount).unwrap(),
                    _ if alph.is_ascii_lowercase() => (97..=122).cycle().nth(alph as usize - 97 + amount).unwrap(),
                    _ => alph as u32,
                })
                .unwrap()
            })
            .collect()
    }

    fn shift_left(&self, amount: usize) -> String {
        self.chars()
            .map(|alph| {
                std::char::from_u32(match alph {
                    _ if alph.is_ascii_uppercase() => (65..= 90).rev().cycle().nth(90  - alph as usize + amount).unwrap(),
                    _ if alph.is_ascii_lowercase() => (97..=122).rev().cycle().nth(122 - alph as usize + amount).unwrap(),
                    _ => alph as u32,
                })
                .unwrap()
            })
            .collect()
    }
}
